### ISSUE TITLE
Context cancel is ignored in Produce

### ISSUE THREAD

**Issue by junsred** [twmb/franz-go#158]
Produce does not return any errors and does not care for context's status when trying to produce to a closed server.

**Comment by twmb** [comment]
What is the code you're using to run into this issue? If I run the following,
```go
package main

import (
	"context"
	"fmt"
	"os"
	"sync"

	"github.com/twmb/franz-go/pkg/kgo"
)

func chk(err error) {
	if err != nil {
		panic(err)
	}
}

func main() {
	cl, err := kgo.NewClient(
		kgo.WithLogger(kgo.BasicLogger(os.Stderr, kgo.LogLevelDebug, nil)),
	)

	chk(err)
	r := &kgo.Record{
		Topic: "foo",
		Value: []byte("hello"),
	}

	chk(cl.Ping(context.Background()))

	ctx, cancel := context.WithCancel(context.Background())
	cancel()

	var wg sync.WaitGroup
	wg.Add(1)
	cl.Produce(ctx, r, func(_ *kgo.Record, err error) {
		fmt.Println(err)
		defer wg.Done()
	})
	wg.Wait()
}
```
with the `foo` topic existing, this is my output:

```
[DEBUG] opening connection to broker; addr: 127.0.0.1:9092, broker: seed 0
[DEBUG] connection opened to broker; addr: 127.0.0.1:9092, broker: seed 0
[DEBUG] issuing api versions request; broker: seed 0, version: 3
[DEBUG] wrote ApiVersions v3; broker: seed 0, bytes_written: 30, write_wait: 37.685µs, time_to_write: 45.955µs, err: <nil>
[DEBUG] read ApiVersions v3; broker: seed 0, bytes_read: 439, read_wait: 122.939µs, time_to_read: 1.979667ms, err: <nil>
[DEBUG] connection initialized successfully; addr: 127.0.0.1:9092, broker: seed 0
[DEBUG] wrote ApiVersions v3; broker: seed 0, bytes_written: 30, write_wait: 2.655544ms, time_to_write: 57.201µs, err: <nil>
[DEBUG] read ApiVersions v3; broker: seed 0, bytes_read: 439, read_wait: 70.008µs, time_to_read: 2.230236ms, err: <nil>
[INFO] immediate metadata update triggered; why: forced load due to unknown produce topic
[INFO] producing to a new topic for the first time, fetching metadata to learn its partitions; topic: foo
[DEBUG] wrote Metadata v11; broker: seed 0, bytes_written: 43, write_wait: 39.982µs, time_to_write: 29.948µs, err: <nil>
[DEBUG] read Metadata v11; broker: seed 0, bytes_read: 652, read_wait: 98.104µs, time_to_read: 2.017946ms, err: <nil>
[INFO] done waiting for metadata for new topic; topic: foo
[INFO] initializing producer id
[DEBUG] opening connection to broker; addr: localhost:9092, broker: 1
[DEBUG] connection opened to broker; addr: localhost:9092, broker: 1
[DEBUG] issuing api versions request; broker: 1, version: 3
[DEBUG] wrote ApiVersions v3; broker: 1, bytes_written: 30, write_wait: 21.22µs, time_to_write: 32.59µs, err: <nil>
[DEBUG] read ApiVersions v3; broker: 1, bytes_read: 439, read_wait: 32.125µs, time_to_read: 2.452929ms, err: <nil>
[DEBUG] connection initialized successfully; addr: localhost:9092, broker: 1
[DEBUG] wrote InitProducerID v4; broker: 1, bytes_written: 34, write_wait: 3.639946ms, time_to_write: 37.027µs, err: <nil>
[DEBUG] read InitProducerID v4; broker: 1, bytes_read: 26, read_wait: 52.599µs, time_to_read: 53.611082ms, err: <nil>
[INFO] producer id initialization success; id: 0, epoch: 0
context canceled
```
(i.e., context canceled is not ignored)

**Comment by junsred** [comment]
```go
package main

import (
	"context"
	"fmt"
	"os"
	"sync"

	"github.com/twmb/franz-go/pkg/kgo"
)

func chk(err error) {
	if err != nil {
		panic(err)
	}
}

func main() {
	cl, err := kgo.NewClient(
		kgo.WithLogger(kgo.BasicLogger(os.Stderr, kgo.LogLevelDebug, nil)),
	)

	chk(err)
	r := &kgo.Record{
		Topic: "foo",
		Value: []byte("hello"),
	}

	//chk(cl.Ping(context.Background()))

	ctx, cancel := context.WithCancel(context.Background())

	var wg sync.WaitGroup
	wg.Add(1)
	cl.Produce(ctx, r, func(_ *kgo.Record, err error) {
		fmt.Println(err)
		defer wg.Done()
	})
	cancel()
	wg.Wait()
}

```

I would expect it to stop but it keeps trying.

**Comment by JacobSMoller** [comment]
@junsred maybe not the solution you want. But for that case I use cl.Ping() to verify that I can connect to the server before doing anything else.

**Comment by junsred** [comment]
@twmb @JacobSMoller is this an issue or something that isn't possible to implement?

**Comment by twmb** [comment]
I'm not sure what the issue you're experiencing is -- the code you sent above comments out the ping in my code. What do the logs say? The code I posted quits the produce immediately.

**Comment by junsred** [comment]
@twmb  Like i said, I would expect the function to be stopped when I call cancel but function gets stuck and halts the program. This is not a desirable case.

**Comment by junsred** [comment]
Trying to produce to a closed server just halts the program without any time outs.

**Comment by twmb** [comment]
Yes—what are your logs? I trust that you are running into the case you
mention, but it cannot be debugged without logs indicating what is
happening inside the client. The code sample I pasted includes a debug
logger. What is the output if you use this debug logger?

**Comment by twmb** [comment]
Also, it sounds like you may be running into the default unlimited retries
aspect of this client. What is the result if you set `RecordRetries` to 1
or something else small?

**Comment by junsred** [comment]
[INFO] producing to a new topic for the first time, fetching metadata to learn its partitions; topic: foo
[INFO] immediate metadata update triggered; why: forced load due to unknown produce topic
[DEBUG] opening connection to broker; addr: 127.0.0.1:9092, broker: seed 0
2022/04/22 14:47:59 context canceled
[WARN] unable to open connection to broker; addr: 127.0.0.1:9092, broker: seed 0, err: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] new topic metadata wait failed, retrying wait; topic: foo, err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] metadata update triggered; why: re-updating metadata due to err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[DEBUG] opening connection to broker; addr: 127.0.0.1:9092, broker: seed 0
[WARN] unable to open connection to broker; addr: 127.0.0.1:9092, broker: seed 0, err: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] new topic metadata wait failed, retrying wait; topic: foo, err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] metadata update triggered; why: re-updating metadata due to err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[DEBUG] opening connection to broker; addr: 127.0.0.1:9092, broker: seed 0
[WARN] unable to open connection to broker; addr: 127.0.0.1:9092, broker: seed 0, err: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] new topic metadata wait failed, retrying wait; topic: foo, err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] metadata update triggered; why: re-updating metadata due to err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[DEBUG] opening connection to broker; addr: 127.0.0.1:9092, broker: seed 0
[WARN] unable to open connection to broker; addr: 127.0.0.1:9092, broker: seed 0, err: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] new topic metadata wait failed, retrying wait; topic: foo, err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] metadata update triggered; why: re-updating metadata due to err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[DEBUG] opening connection to broker; addr: 127.0.0.1:9092, broker: seed 0
[WARN] unable to open connection to broker; addr: 127.0.0.1:9092, broker: seed 0, err: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] new topic metadata wait failed, retrying wait; topic: foo, err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] metadata update triggered; why: re-updating metadata due to err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[DEBUG] opening connection to broker; addr: 127.0.0.1:9092, broker: seed 0
[WARN] unable to open connection to broker; addr: 127.0.0.1:9092, broker: seed 0, err: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] new topic metadata wait failed, retrying wait; topic: foo, err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] metadata update triggered; why: re-updating metadata due to err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[DEBUG] opening connection to broker; addr: 127.0.0.1:9092, broker: seed 0
[WARN] unable to open connection to broker; addr: 127.0.0.1:9092, broker: seed 0, err: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] new topic metadata wait failed, retrying wait; topic: foo, err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] metadata update triggered; why: re-updating metadata due to err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[DEBUG] opening connection to broker; addr: 127.0.0.1:9092, broker: seed 0
[WARN] unable to open connection to broker; addr: 127.0.0.1:9092, broker: seed 0, err: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] new topic metadata wait failed, retrying wait; topic: foo, err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] metadata update triggered; why: re-updating metadata due to err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[DEBUG] opening connection to broker; addr: 127.0.0.1:9092, broker: seed 0
[WARN] unable to open connection to broker; addr: 127.0.0.1:9092, broker: seed 0, err: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] new topic metadata wait failed, retrying wait; topic: foo, err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] metadata update triggered; why: re-updating metadata due to err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[DEBUG] opening connection to broker; addr: 127.0.0.1:9092, broker: seed 0
[WARN] unable to open connection to broker; addr: 127.0.0.1:9092, broker: seed 0, err: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] new topic metadata wait failed, retrying wait; topic: foo, err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] metadata update triggered; why: re-updating metadata due to err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[DEBUG] opening connection to broker; addr: 127.0.0.1:9092, broker: seed 0
[WARN] unable to open connection to broker; addr: 127.0.0.1:9092, broker: seed 0, err: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
[INFO] new topic metadata wait failed, retrying wait; topic: foo, err: unable to dial: dial tcp 127.0.0.1:9092: connectex: Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı.
... continues like this even though context is cancelled

**Comment by junsred** [comment]
kgo.RecordRetries(1)
and
kgo.RecordDeliveryTimeout(time.Second*5),

Does seem to work but i thought the option for it was ProduceRequestTimeout not RecordDeliveryTimeout which confused me. Maybe RecordDeliveryTimeout or RecordRetries should be added to examples?

**Comment by junsred** [comment]
@twmb

**Comment by twmb** [comment]
@junsred `ProduceRequestTimeout` is a bit of a leaky abstraction: the `ProduceRequest` type itself has a `TimeoutMillis` field that instructs the _broker_ to quit handling a request once the time has elapsed. It doesn't mean anything about individual record time processing.

I'd be open for an example being added using `RecordDeliveryTimeout` (which I recommend over `RecordRetries`.

Also FWIW: as of Kafka 3.0, the default setting for Java clients is to "[enable the strongest delivery guarantee](https://cwiki.apache.org/confluence/display/KAFKA/KIP-679%3A+Producer+will+enable+the+strongest+delivery+guarantee+by+default)", meaning unlimited retries.

I don't think I'd want to change the client to automatically fail records if the cluster is down by default, though. That may result in some surprise if a person sets up a client and then the cluster is down for a blip of time and they have the client to have unlimited retries.

**Comment by junsred** [comment]
@twmb how about context's status? It should at least consider context's status before trying to connect again.

**Comment by twmb** [comment]
I can add a new select in this block: https://github.com/twmb/franz-go/blob/01d0c45d69d274d4fab20ab16f469d380f929462/pkg/kgo/producer.go#L818

If the first context for unknown records is canceled, all records buffered for the unknown topic will be failed.

This block also needs to take into account if the client is aborting, although that error is slightly less important considering the block will fail when either (1) record retries are hit, (2) record timeout is hit, (3) record context is canceled, or [critically] (4) the topic failed to load 4 times (which happens within seconds if the cluster is reachable).

**Comment by twmb** [comment]
This was closed in 42e5b57, which will be tagged in v1.5 soon (I'm aiming for Monday afternoon).

**Comment by twmb** [comment]
This is now released in v1.5.

**Comment by twmb** [comment]
What are your logs?
